feat: add consistently interactive audit

patrickhulce opened the pull request:
https://docs.google.com/document/d/1GGiI9-7KeY3TPqS3YT271upUVimo-XiL5mwWorDUD4c/preview

patrickhulce commented:
PTAL :)

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+
+    const quietPeriods = [];
+    longTasks.forEach((task, index) => {
+      if (index === 0) {
```
hey a forEach! 😁   😍

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+    let inflight = 0;
+    let quietPeriodStart = 0;
+    const quietPeriods = [];
+    timeBoundaries.forEach(boundary => {
```
hey a forEach! 😁   😍

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+   */
+  static _findNetworkQuietPeriods(networkRecords, traceOfTab) {
+    const traceEnd = traceOfTab.timestamps.traceEnd;
+    const timeBoundaries = networkRecords.reduce((boundaries, record) => {
```
can you add a comment to clarify these timeBoundaries are just the timestamps of when requests start and end

patrickhulce replied:
done

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+    const quietPeriods = [];
+    timeBoundaries.forEach(boundary => {
+      if (boundary.isStart) {
+        // we are exiting a quiet period
```
// we've just started a new request. are we exiting a quiet period?

patrickhulce replied:
done

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+        inflight++;
+      } else {
+        inflight--;
+        // we are entering a quiet period
```
// we've just completed a request. are we entering a new quiet period?

patrickhulce replied:
done

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+      return boundaries;
+    }, []).sort((a, b) => a.time - b.time);
+
+    let inflight = 0;
```
s/inflight/inflightReqs/

patrickhulce replied:
done-ish :)

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+    const FMPTsInMs = traceOfTab.timestamps.firstMeaningfulPaint;
+
+    const networkQuietPeriods = this._findNetworkQuietPeriods(networkRecords, traceOfTab)
+        .filter(period => period.end > FMPTsInMs + REQUIRED_QUIET_WINDOW &&
```
extract inner fn to const?

patrickhulce replied:
done

paulirish commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,242 @@
+        .filter(period => period.end > FMPTsInMs + REQUIRED_QUIET_WINDOW &&
+            period.end - period.start >= REQUIRED_QUIET_WINDOW);
+    const cpuQuietPeriods = this._findCPUQuietPeriods(longTasks, traceOfTab)
+        .filter(period => period.end > FMPTsInMs + REQUIRED_QUIET_WINDOW &&
```
and reuse here

brendankenny commented:
> should this be in trace-of-tab?

er, that would be `pwmetrics-events` :)

patrickhulce commented:
> should this be in pwmetrics-events?

yep, added 👍 

> if this is what we're going to be pushing at I/O, should the various audits switching over to firstInteractive from TTI in #2013 be switching to this instead?

sadly, yes. I replaced it with firstInteractive since it's most comparable to our previous definition in spirit compared to the pessimism of this one, but we'll probably want to switch the load fast enough for PWA one sometime before I/O if that policy sticks, it has unfortunate effects for other audits like the unused images since by definition consistently interactive would wait for the images to finish, so I'm not sold on switching everything just yet...

patrickhulce commented:
rebased and ready to take on the day 🌥  -> 🌤

patrickhulce commented:
waiting for #2062 to add a real trace test

brendankenny commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+    const timeBoundaries = [];
+    networkRecords.forEach(record => {
+      const scheme = record.parsedURL && record.parsedURL.scheme;
+      if (scheme === 'data' || scheme === 'ws') {
```
might be overkill, but should we convert this to a `ignoredNetworkSchemes` array and pull it into a constant?

patrickhulce replied:
sure, done

brendankenny commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+          throw new Error('No firstMeaningfulPaint found in trace.');
+        }
+
+        const longTasks = TracingProcessor.getMainThreadTopLevelEvents(traceModel, trace)
```
though I guess `_findNetworkQuietPeriods` would then also need to find out and clip by `FMP`

patrickhulce replied:
wouldn't really be able to remove anything since we still need to check that the period is 5s after FMP (even if long tasks are ignored before FMP) so it's just the saved long tasks we're iterating through, and IMO it makes the logic harder to parse that FMP has already been dealt with in CPU but we still have to deal with it in network, fine leaving?

brendankenny replied:
> fine leaving?

yeah, I guess it's just awkward one way or the other. SG

brendankenny commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+          traceOfTab.timestamps.domContentLoaded
+        ) * 1000;
+        const timeInMs = (timestamp - traceOfTab.timestamps.navigationStart * 1000) / 1000;
+        const extendedInfo = Object.assign(quietPeriodInfo, {timestamp, timeInMs});
```
shape of `quietPeriodInfo` is difficult to track down. Can it be made more evident here?

patrickhulce replied:
added typedef to `findOverlappingQuietPeriods` 👍

brendankenny replied:
that is...not a terribly useful type definition :P If I wanted to use that `extendedInfo` I would have to check the `param` signatures of `_findCPUQuietPeriods` and `_findNetworkQuietPeriods` and read `findOverlappingQuietPeriods` to make sure there's only one code escape path. We don't have a great place for documenting artifact shapes, but it would be nice to always do it, even if in an awkward place

patrickhulce replied:
added a `TimePeriod` typedef

```
/**
 * @typedef {{
 *     start: number,
 *     end: number,
 * }}
 */
let TimePeriod; // eslint-disable-line no-unused-vars
```

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+//   https://www.desmos.com/calculator/uti67afozh
+const SCORING_POINT_OF_DIMINISHING_RETURNS = 1700;
+const SCORING_MEDIAN = 10000;
+const SCORING_TARGET = 5000;
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+      description: 'Consistently Interactive (beta)',
+      helpText: 'The point at which most network resources have finished loading and the ' +
+          'CPU is idle for a prolonged period.',
+      optimalValue: SCORING_TARGET.toLocaleString() + 'ms',
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  /**
+   * @param {!Array} networkRecords
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+      } else {
+        numInflightRequests--;
+        // we've just completed a request. are we entering a quiet period?
+        if (numInflightRequests <= ALLOWED_CONCURRENT_REQUESTS) {
```
probably not? I can't think of one, so done :)

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+    });
+
+    // Check if the trace ended in a quiet period
+    if (quietPeriodStart !== Infinity) {
```
yep sgtm, this func was mostly just the network throughput computation modified to line up with [deep's implementation](https://codereview.chromium.org/2610183003/patch/140001/150003), but it could be clearer 👍

patrickhulce replied:
come to think of it, in this version the condition will never be true anyway...I'll have to look into how end times are handled

patrickhulce replied:
yikes, we've actually got a problem with network recorder :/ it only saves records on finish

brendankenny replied:
hmm, good catch. That was for expedience at the time, but that was a very long time ago and we could probably do with some more nuanced data.

A quick fix would be to just look directly at the `devtoolsLog` for unclosed requests, but what are you thinking?

patrickhulce replied:
just fixed, adding the request on start instead and checking `.finished`

brendankenny replied:
> just fixed, adding the request on start instead and checking `.finished`

this is the right solution, but nothing else that depends on network records is checking for this as they all could assume that all requests were finished when they were written. They're often just taking all of them and filtering on whatever they need to look for. We need to do a survey of all the uses, fix where it makes sense, add unit tests like you added in a5d62b8 for CI, etc

patrickhulce replied:
I *think* I've addressed all the usages here now, some places were already checking for it

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number}>} longTasks
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number}>} longTasks
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+
+  /**
+   * @param {!Array<{start: number, end: number}>} longTasks
+   * @param {{timestamps: {navigationStart: number, firstMeaningfulPaint: number,
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+   * @param {!Array<{start: number, end: number}>} longTasks
+   * @param {{timestamps: {navigationStart: number, firstMeaningfulPaint: number,
+   *    traceEnd: number}}} traceOfTab
+   * @return {!Object}
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+        .filter(isLongEnoughQuietPeriod);
+
+    const cpuQueue = cpuQuietPeriods.slice();
+    const networkQueue = networkQuietPeriods.slice();
```
they're also returned in extended info

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,248 @@
+    let networkCandidate = networkQueue.shift();
+    while (cpuCandidate && networkCandidate) {
+      if (cpuCandidate.start >= networkCandidate.start) {
+        // CPU starts later than network, it must be contained by network or we check the next network
```
done

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+
+    return Promise.all(computedTraceArtifacts)
+      .then(([traceModel, traceOfTab]) => {
+        if (!traceOfTab.timestamps.firstMeaningfulPaint) {
```
done

patrickhulce replied:
(though in the spec I'm not actually sure this is true, it just seemed like the proper thing to do. I should raise it)

brendankenny replied:
> (though in the spec I'm not actually sure this is true, it just seemed like the proper thing to do. I should raise it)

wouldn't no FMP be a bug in Chrome/Lighthouse? Unless you mean something else

patrickhulce replied:
I was talking about the fact that we're doing the max of DCL and the value. I wasn't sure that's in the spec or if I added that to be sure it doesn't happen before TTFI

patrickhulce commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,247 @@
+        }
+
+        const longTasks = TracingProcessor.getMainThreadTopLevelEvents(traceModel, trace)
+            .filter(event => event.end - event.start >= 50);
```
done

paulirish commented:
I want this guy in ASAP, but not going to rush it. :)
@brendankenny your move here.

patrickhulce commented:
FYI @brendankenny I split that part out into a separate PR so it's easier to review, PTAL #2197

brendankenny commented:
> FYI @brendankenny I split that part out into a separate PR

🎉

patrickhulce commented:
rebased and ready to go 💥

brendankenny commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,258 @@
+  /**
+   * Finds all time periods where the number of inflight requests is less than or equal to the
+   * number of allowed concurrent requests (2).
+   * @param {!Array<WebInspector.NetworkRequest>} networkRecords
```
nit: `!Array<!WebInspector.NetworkRequest>`

patrickhulce replied:
done

brendankenny commented on lighthouse-core/audits/consistently-interactive.js:
```diff
@@ -0,0 +1,258 @@
+  /**
+   * Finds the first time period where a network quiet period and a CPU quiet period overlap.
+   * @param {!Array<{start: number, end: number}>} longTasks
+   * @param {!Array<WebInspector.NetworkRequest>} networkRecords
```
`!Array<!WebInspector.NetworkRequest>`

patrickhulce replied:
done
